### PR TITLE
add poetry support

### PR DESCRIPTION
**Pull request by An-Li-magicloud**
Changes:
* Poetry adoption
* release.yml workflow change in poetry way(not tested)
*  ``tidy3d configure`` cli for apikey setup
    * install in from local codebase: ``poetry install`` 
* Add apikey as new authentication way, auth order:
    *  SIMCLOUD_APIKEY environment variable
    * ~/.tidy3d/config for api key
    * credential

### CODE REVIEW

**Review by tylerflex — COMMENTED** [review]
Hey @An-Li-magicloud just a comment: I noticed that `tox.ini` was removed and now the tests just run `pytest`. Is there any way to keep `tox`? It is very useful for us to test many python versions 3.7+, OS, and configure things like environment variables, which we need for some features coming soon. If there is a way to keep it, that would be ideal. This article describes a way to do it [link](https://levelup.gitconnected.com/python-a-toxic-work-environment-tox-poetry-b8200563c474).
`tox.ini`
```
[tox]
skipsdist = true
envlist = py37, py38

[testenv]
whitelist_externals = poetry
commands =
    poetry install -v
    poetry run pytest
```

**Comment by tylerflex** [comment]
Ah good catch, thanks for correcting that. This is great, let me know when you think it is ready for me to test and review. Probably best way is to assign me as a reviewer to this PR and leave a note explaining how to use. Thanks!

File: tidy3d/web/httputils.py [hunk 1]
```diff
@@ -1,15 +1,61 @@
 """ handles communication with server """
+import os
+
 # import os
```
Comment (tylerflex) [inline tidy3d/web/httputils.py:4]: just uncomment this and remove previous lines

File: tidy3d/web/httputils.py [hunk 2]
```diff
@@ -1,15 +1,61 @@
...
+
+
+def api_key():
+    """
+    Get the api key for the current environment.
+    :return:
+    """
+    if os.environ.get(SIMCLOUD_APIKEY):
+        return os.environ.get(SIMCLOUD_APIKEY)
+    if os.path.exists(f"{expanduser('~')}/.tidy3d/config"):
```
Comment (tylerflex) [inline tidy3d/web/httputils.py:29]: can we define something like
```py
TIDY3D_DIR = f"{expanduser('~')}/.tidy3d
CONFIG_FILE = TIDY3D_DIR + "/config"
```
as constants in this module and then use them below and in the CLI app? so if we change the file path or name some time in the future, we can do it in one place.

**Review by tylerflex — CHANGES_REQUESTED** [review]
Hey An, thanks for this, it's a really good start. I was able to use poetry to install tidy3d successfully, but had a few comments that we can address for now. A few are more organizational, but then I ran into an error when trying to run a simulation through the webapi using the current configuration:
* We probably need a tutorial / guide / documentation page for how to set up each way to authenticate and also a brief explanation of how to use poetry for tidy3d. Eventually, we want this to go in our documentation, for example a markdown or RST file, perhaps for now just some tutorial in `web/README.md` file is sufficient until we figure out how to include in the docs.
* Seems like only python < 3.11 is supported. Can we make it work for 3.11 as well?
* the cli should probably not go in `plugins/cli/`, I think it should be in `web/cli`. `plugins` are supposed to be simulation features built around the core `components`. As part of this, `test_cli.py` should also go in `tests/test_web/` and there are a few other changes.
* I tried to test the authentication (without API key) with the following (clean environment, from `tidy3d/` root directory, not using API key, instead saved credentials like the `develop` branch authentication) and got an error:
```
poetry install
poetry run python
>>> import tidy3d.web as web
>>> sim = td.Simulation(grid_spec=td.GridSpec(wavelength=1.0), size=(1,1,1), run_time=1e-12)
[09:30:45] WARNING  No sources in simulation.         simulation.py:464
>>> web.run(sim, task_name='poetry test')
[09:30:54] INFO     Using Tidy3D credentials from stored     auth.py:70
                    file.
[09:30:56] INFO     Authentication successful.               auth.py:30
[09:30:57] INFO     Authentication successful.               auth.py:30
[09:30:59] INFO     Authentication successful.               auth.py:30
           ERROR    {"timestamp":"2022-12-21T15:30:59.693+00: log.py:33
                    00","status":401,"error":"Unauthorized","
                    message":"Unauthorized:None account
                    associate with this API
                    Key.","path":"/tidy3d/project"}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/twhughes/Documents/Flexcompute/tidy3d-docs/tidy3d/tidy3d/web/webapi.py", line 64, in run
    task_id = upload(
  File "/Users/twhughes/Documents/Flexcompute/tidy3d-docs/tidy3d/tidy3d/web/webapi.py", line 109, in upload
    folder = _query_or_create_folder(folder_name)
  File "/Users/twhughes/Documents/Flexcompute/tidy3d-docs/tidy3d/tidy3d/web/webapi.py", line 557, in _query_or_create_folder
    raise e
  File "/Users/twhughes/Documents/Flexcompute/tidy3d-docs/tidy3d/tidy3d/web/webapi.py", line 550, in _query_or_create_folder
    resp = http.get(method)
  File "/Users/twhughes/Documents/Flexcompute/tidy3d-docs/tidy3d/tidy3d/web/httputils.py", line 87, in wrapper
    raise WebError(resp.text)
tidy3d.log.WebError: (WebError(...), '{"timestamp":"2022-12-21T15:30:59.693+00:00","status":401,"error":"Unauthorized","message":"Unauthorized:None account associate with this API Key.","path":"/tidy3d/project"}')
```
* Where can I get an API key to test?

**Comment by An-Li-magicloud** [comment]
> Hey An, thanks for this, it's a really good start. I was able to use poetry to install tidy3d successfully, but had a few comments that we can address for now. A few are more organizational, but then I ran into an error when trying to run a simulation through the webapi using the current configuration:
> 
> * We probably need a tutorial / guide / documentation page for how to set up each way to authenticate and also a brief explanation of how to use poetry for tidy3d. Eventually, we want this to go in our documentation, for example a markdown or RST file, perhaps for now just some tutorial in `web/README.md` file is sufficient until we figure out how to include in the docs.
> * Seems like only python < 3.11 is supported. Can we make it work for 3.11 as well?
> * the cli should probably not go in `plugins/cli/`, I think it should be in `web/cli`. `plugins` are supposed to be simulation features built around the core `components`. As part of this, `test_cli.py` should also go in `tests/test_web/` and there are a few other changes.
> * I tried to test the authentication (without API key) with the following (clean environment, from `tidy3d/` root directory, not using API key, instead saved credentials like the `develop` branch authentication) and got an error:
> 
> ```
> poetry install
> poetry run python
> >>> import tidy3d.web as web
> >>> sim = td.Simulation(grid_spec=td.GridSpec(wavelength=1.0), size=(1,1,1), run_time=1e-12)
> [09:30:45] WARNING  No sources in simulation.         simulation.py:464
> >>> web.run(sim, task_name='poetry test')
> [09:30:54] INFO     Using Tidy3D credentials from stored     auth.py:70
>                     file.
> [09:30:56] INFO     Authentication successful.               auth.py:30
> [09:30:57] INFO     Authentication successful.               auth.py:30
> [09:30:59] INFO     Authentication successful.               auth.py:30
>            ERROR    {"timestamp":"2022-12-21T15:30:59.693+00: log.py:33
>                     00","status":401,"error":"Unauthorized","
>                     message":"Unauthorized:None account
>                     associate with this API
>                     Key.","path":"/tidy3d/project"}
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/Users/twhughes/Documents/Flexcompute/tidy3d-docs/tidy3d/tidy3d/web/webapi.py", line 64, in run
>     task_id = upload(
>   File "/Users/twhughes/Documents/Flexcompute/tidy3d-docs/tidy3d/tidy3d/web/webapi.py", line 109, in upload
>     folder = _query_or_create_folder(folder_name)
>   File "/Users/twhughes/Documents/Flexcompute/tidy3d-docs/tidy3d/tidy3d/web/webapi.py", line 557, in _query_or_create_folder
>     raise e
>   File "/Users/twhughes/Documents/Flexcompute/tidy3d-docs/tidy3d/tidy3d/web/webapi.py", line 550, in _query_or_create_folder
>     resp = http.get(method)
>   File "/Users/twhughes/Documents/Flexcompute/tidy3d-docs/tidy3d/tidy3d/web/httputils.py", line 87, in wrapper
>     raise WebError(resp.text)
> tidy3d.log.WebError: (WebError(...), '{"timestamp":"2022-12-21T15:30:59.693+00:00","status":401,"error":"Unauthorized","message":"Unauthorized:None account associate with this API Key.","path":"/tidy3d/project"}')
> ```
> 
> * Where can I get an API key to test?

3.11 support is added, but due to h5py doesn't have a prebuild binary for 3.11 which lead to error during depedencies install, I didn't enable it in github action.

[x] readme.md is added.

File: tidy3d/web/httputils.py [hunk 3]
```diff
@@ -1,15 +1,59 @@
...
```
Comment (tylerflex) [inline tidy3d/web/httputils.py:27]: need to `from .cli.app import CONFIG_FILE` and use it here and int he line below?

Comment (An-Li-magicloud) [reply tidy3d/web/httputils.py]: I think it would be better to have creator manage the constant, in this case, we could leave it in ``app.py``

Comment (tylerflex) [reply tidy3d/web/httputils.py]: agree

File: tidy3d/web/cli/app.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        Set the authentication.
+        :param req:
+        :return:
+        """
+        req.headers["simcloud-api-key"] = apikey
+        return req
+
+    resp = requests.get(f"{DEFAULT_CONFIG.web_api_endpoint}/apikey", auth=auth)
+    if resp.status_code == 200:
+        click.echo("Configure successfully.")
```
Comment (tylerflex) [inline tidy3d/web/cli/app.py:59]: "Configured successfully."

File: tidy3d/web/cli/app.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    Tidy3d command line tool.
+    """
+
+
+@click.command()
+@click.option(
+    "--apikey", prompt=config_description if "config_description" in globals() else "API Key"
+)
+def configure(apikey):
+    """
```
Comment (tylerflex) [inline tidy3d/web/cli/app.py:35]: We use [numpy style docstrings](https://numpydoc.readthedocs.io/en/latest/format.html) in Tidy3d. For consistency, can we change these and the other ones to numpy style? [Example](https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/web/webapi.py#L78-L100)

File: tidy3d/web/cli/readme.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+You can find your API key in the web http://tidy3d.simulation.cloud
+
+
+### Setup API key
+
+#### Environment Variable
+``export SIMCLOUD_API_KEY="your_api_key"``
+
+#### Command line
+``tidy3d configure``, then enter your API key
```
Comment (tylerflex) [inline tidy3d/web/cli/readme.md:14]: "then enter your API key when prompted"

Comment (tylerflex) [inline tidy3d/web/cli/readme.md:14]: Quick question, I needed to do something like 
```
poetry run tidy3d configure
```
for this to work. How did you set up poetry when testing? Did you use `poetry shell` first to start the virtual env and then run the other commands? `pytest tests` etc?

**Review by tylerflex — CHANGES_REQUESTED** [review]

File: tidy3d/web/cli/readme.md (hunk 6 above)
Comment (An-Li-magicloud) [reply tidy3d/web/cli/readme.md:14]: Yes, poetry recommend to use ``poetry shell`` to enter virtual env during developing.

File: tox.ini [hunk 7]
```diff
@@ -56,4 +52,4 @@ commands =
     pytest -rA tests/test_plugins/test_resonance_finder.py
     pytest -rA tests/test_web/test_auth.py
     pytest -rA tests/test_web/test_task.py
```
Comment (tylerflex) [inline tox.ini:54]: Also, we should add `tests/test_web/test_cli.py`

**Review by tylerflex — CHANGES_REQUESTED** [review]

File: .github/workflows/run_tests.yml [hunk 8]
```diff
@@ -17,14 +17,17 @@ jobs:
         platform: [ubuntu-latest, macos-latest, windows-latest]
```
Comment (tylerflex) [inline .github/workflows/run_tests.yml:17]: Should we add `3.11` to `python-version`?

File: pyproject.toml [hunk 9]
```diff
@@ -1,9 +1,65 @@
...
+readme = "README.md"
+packages = [
+    { include = "tidy3d" },
+    { include = "tidy3d/web" },
+    { include = "tidy3d/plugins" },
 ]
-build-backend = "setuptools.build_meta"
+
+
+[tool.poetry.dependencies]
```
Comment (tylerflex) [inline pyproject.toml:14]: I feel like we need more granular control over the dependencies like we have currently.

* `core` core packages, could be installed with `poetry install --only core`
* `web` things needed for web only
* `basic` the default install, `core` + `web` + `scipy` should be installed by default.
* `dev`should be optional, ie `poetry install --with dev`

[More info on dependency groups](https://python-poetry.org/docs/managing-dependencies/#optional-groups)
[Our current dependencies](https://github.com/flexcompute/tidy3d/tree/develop/requirements)

File: .github/workflows/run_tests.yml (hunk 8 above)
Comment (An-Li-magicloud) [reply .github/workflows/run_tests.yml:17]: h5py doesn't have prebuild for 3.11 which leads python to compile it and cause error. I would recommend to wait h5py upload precompile binary to pypi.

File: pyproject.toml (hunk 9 above)
Comment (An-Li-magicloud) [reply pyproject.toml:14]: It seems poetry doesn't support one group declare other groups as dependencies. We maybe thinking of an other orchestrate.

**Comment by tylerflex** [comment]
Note: given some discussions in #642 and #643 , we may want to set shapely == ^2.0.0

**Comment by tylerflex** [comment]
I'm reluctant to merge this because more and more we are relying on dependency groups. For example, we have added `[jax]` as a group and perhaps soon `[gdstk]`. If we can't figure out a good solution to this, I think we should stick with the `setup.py` solution as the main point of using poetry initially was for the CLI, right?

**Comment by An-Li-magicloud** [comment]
Sorry for the late response. Here is another solution by using ``extras``, here is an example to make gdstk optional:

```toml
[tool.poetry.dependencies]
gdstk = { version = "xxx", optional=true }

[tool.poetry.extras]
gds = ["gdstk"]
```
When user need gdstk in there local environment, use ``-E `` to declare deps.

```
poetry install -E gds
```
https://python-poetry.org/docs/pyproject/#extras

**Comment by tylerflex** [comment]
Thanks for looking into this An, I'm planning to merge the webapi refactor next (#730) and then revisiting this once that is in.

I think for me it seems we can do everything we currently need already with setup.py. If we want to switch to poetry, it will involve some learning curve and adjustment to the way we do things, including on the backend, so I think we want to make sure it is worth the investment if we do change, especially now that we are trying to finish a few things before a major release.

In your opinion, what would be the main advantage(s) of moving to poetry? Do you think it's worth it for the future of Tidy3D?

**Comment by An-Li-magicloud** [comment]
Here is what I felt poetry is better than raw pip:
* virtual env integrated
* Unified environment -- we want to solve dependencies once, then run it in every environments, include our local development, user's local machine and servers.
* Security/deprecated upgrade. Poetry could fine-grained manage dependencies, it  frees users from upgrading hell when they use legacy  python versions, eg. 3.6, 3.7, and pynum and it's related packages compliant each other.
```
matplotlib = [{ python = "~3.7", version = "~3.5" }, { python = "^3.8", version = "^3.6" }]
pandas = [{ python = "~3.7", version = "^1.3.0"}, {python = "^3.8", version = "^1.5.0"}]
```

**Comment by tylerflex** [comment]
I will close this PR for now because it's diverged so much from the main branch. But when (and if) we do poetry, I will pull aspects from this PR and maybe make a new one.

**Pull request closed without merging**